src/bin/edit/localization.rs

```diff
@@ -1,132 +1,132 @@
 use edit::sys;
 
-macro_rules! first_item {
-    ($first:tt $($rest:tt)*) => {
-        $first
+macro_rules! replace_expr {
+    ($_t:tt $sub:expr) => {
+        $sub
     };
 }
+macro_rules! count_sizes {
+   ($($tts:tt)*) => {0usize $(+ replace_expr!($tts 1usize))*};
+}
 
-macro_rules! gen_l18n {
+macro_rules! gen_l10n {
     (
-    $($lang_name_1:ident),+;
-    $($string_name:ident: {
-    	$($lang_name_2:ident: $string:expr),+$(,)?
-	}),+$(,)?) => {
-		  #[cfg(test)]
-		  #[derive(Debug, Clone, Copy, PartialEq, Eq)]
-        pub enum LocId {
         $(
-            $string_name
-        ),+
-        ,Count,
-        }
-
-
-        #[cfg(not(test))]
-		  #[derive(Clone, Copy, PartialEq, Eq)]
-         pub enum LocId {
-         $(
-               $string_name
-         ),+
-         ,Count,
-         }
-
+            $lang_name_str:literal:
+            $lang_name_1:ident $(: [$($prefix:literal),+])?
+        ),+;
+        $(
+            $(#[doc = $string_id_doc:literal])?
+            $string_name:ident: {
+                $($lang_name_2:ident: $string:expr),+$(,)?
+            }
+        ),+$(,)?
+    ) => {
+        const STRING_COUNT: usize = count_sizes!($($string_name)+);
+        const LANG_COUNT: usize = count_sizes!($($lang_name_1)+);
 
+        #[derive(Clone, Copy, PartialEq, Eq)]
+        #[cfg_attr(test, derive(Debug))]
+        pub enum LocId {
+            $(
+                $(#[doc = $string_id_doc])?
+                $string_name,
+            )+
+        }
 
         #[allow(non_camel_case_types)]
         #[derive(Clone, Copy, PartialEq, Eq)]
+        #[cfg_attr(test, derive(Debug))]
         pub enum LangId {
-        		$(
-          $lang_name_1
-          ),+
-            ,Count,
+            $(
+                #[doc = $lang_name_str]
+                $lang_name_1,
+            )+
         }
         impl LangId {
-            const fn fallback() -> Self {
-          		use LangId::*;
-          		first_item!($($lang_name_1)+)
+            pub const fn first() -> Self {
+                Self::map()[0].1
+            }
+            pub const fn map() -> &'static [(&'static str, Self)] {
+                use LangId::*;
+                &[
+                    $(
+                        (stringify!($lang_name_1), $lang_name_1),
+                    )+
+                ]
             }
-            fn as_str(&self) -> &'static str {
-          		use LangId::*;
-            	match self {
- 		        		$(
-		             	$lang_name_1 => stringify!($lang_name_1),
-		             )+
-						Count => "Count"
-	             }
+            pub const fn prefix_map() -> &'static [(&'static str, Self)] {
+                use LangId::*;
+                &[
+                    $(
+                        (stringify!($lang_name_1), $lang_name_1),
+                        $($(($prefix, $lang_name_1),)+)?
+                    )+
+                ]
+            }
+            pub const fn name(&self) -> &'static str {
```

Comment:
This would be for using the name of the language as a string in the UI. For example if a language switcher is added.